langkit/documentation.py

```diff
@@ -248,6 +248,14 @@ def instantiate_templates(doc_dict: Dict[str, str]) -> DocDatabase:
         Raised when introspection functions (``${ctx.lib_name}.Introspection``)
         are improperly used to evaluate a node data (field or property).
     """,
+    'langkit.introspection.bad_type_error': """
```

Comment:
The fact of having a `Bad_Type_Error` doesn't seem very consistent with the fact that:

1. Access to fields on the wrong type will raise a `Node_Data_Evaluation_Error`:

```
         (raise Node_Data_Evaluation_Error with "no such field on this node");
```

2. Querying the `Index` of a field on the wrong node type will raise a `Constraint_Error`

3. The fact that type constraints are called `Value_Constraint`s.

TBH I would much rather type constraints are called `Type_Constraint` and the two case above raise a `Bad_Type_Error`, that would be much easier to understand and deal with, the current design is inconsistent IMO.

EDIT: Specifically for the case of `Constraint_Error`, every error here on `Value_Type` could be a constraint error, but we'd rather not do that for general design reasons, so I think we should just not use it.

Reply:
As discussed live: I’ll merge `Node_Data_Evaluation_Error` and `Invalid_Field` into `Bad_Type_Error`. In addition, I’ll replace most `Constraint_Error` with `Bad_Type_Error` too. And let’s go for the `Value_Constraint` renaming.